src/remote/remote.go

```diff
@@ -0,0 +1,326 @@
+// (https://github.com/bazelbuild/remote-apis) which Please can use to distribute
+// work to remote servers.
+package remote
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/bazelbuild/remote-apis/build/bazel/semver"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	bs "google.golang.org/genproto/googleapis/bytestream"
+	"google.golang.org/grpc"
+	_ "google.golang.org/grpc/encoding/gzip" // Registers the gzip compressor at init
+	"gopkg.in/op/go-logging.v1"
+
+	"github.com/thought-machine/please/src/core"
+	"github.com/thought-machine/please/src/fs"
+)
+
+var log = logging.MustGetLogger("remote")
+
+// Timeout to initially contact the server.
+const dialTimeout = 5 * time.Second
+
+// Timeout for actual requests
+const reqTimeout = 2 * time.Minute
+
+// Maximum number of times we retry a request.
+const maxRetries = 3
+
+// The API version we support.
+var apiVersion = semver.SemVer{Major: 2}
+
+// A Client is the interface to the remote API.
+//
+// It provides a higher-level interface over the specific RPCs available.
+type Client struct {
+	actionCacheClient pb.ActionCacheClient
+	storageClient     pb.ContentAddressableStorageClient
+	bsClient          bs.ByteStreamClient
+	initOnce          sync.Once
+	state             *core.BuildState
+	err               error // for initialisation
+
+	// This is for servers have have multiple instances. Right now we never set it but
+	// we keep this here to remind us where it would need to go in the API.
+	instance string
+
+	// Server-sent cache properties
+	maxBlobBatchSize int64
+	cacheWritable    bool
+
+	// Cache this for later
+	bashPath string
+}
+
+// instance is the singleton client instance for Get()
+var instance *Client
+
+// New returns a new Client instance.
+// It begins the process of contacting the remote server but does not wait for it.
+func New(state *core.BuildState) *Client {
+	c := &Client{state: state}
+	go c.CheckInitialised() // Kick off init now, but we don't have to wait for it.
+	return c
+}
+
+// Get is like New but it populates and retrieves a singleton instance.
+func Get(state *core.BuildState) *Client {
+	if instance == nil {
+		instance = New(state)
+	}
+	return instance
+}
+
+// CheckInitialised checks that the client has connected to the server correctly.
+func (c *Client) CheckInitialised() error {
+	c.initOnce.Do(c.init)
+	return c.err
+}
+
+// init is passed to the sync.Once to do the actual initialisation.
+func (c *Client) init() {
+	c.err = func() error {
+		// TODO(peterebden): We may need to add the ability to have multiple URLs which we
+		//                   would then query for capabilities to discover which is which.
+		conn, err := grpc.Dial(c.state.Config.Remote.URL.String(),
+			grpc.WithTimeout(dialTimeout),
+			grpc.WithInsecure(),
```

Comment:
Shouldn't this depend on the URL passed to it?

Reply:
no, grpc urls typically don't have a protocol so we can't infer it. I'll add a TODO, it really needs to be specified in the config (I expect we will want the ability to pass a CA at least for TLS).